patches update_h2 script to remove OS checks

diff -urN hexen2source-1.3.0/gamecode/pak_v111/update_h2 hexen2source-1.3.0x/gamecode/pak_v111/update_h2
--- hexen2source-1.3.0/gamecode/pak_v111/update_h2	2005-08-29 11:49:23.000000000 +0300
+++ hexen2source-1.3.0x/gamecode/pak_v111/update_h2	2005-08-29 13:18:22.000000000 +0300
@@ -1,16 +1,5 @@
 #!/bin/sh
 
-os=`uname -s`
-
-if [[ ! -f loki_patch.$os ]]; then
-	echo ""
-	echo "A loki_patch binary for your $os operating system cannot"
-	echo "be found. Please compile one from the sources and rename"
-	echo "to loki_patch.$os for this script to use it. Exitting..."
-	echo ""
-	exit 1
-fi
-
 echo ""
 echo "=============================================================="
 echo "This will patch Hexen2 PAK files (cdrom-1.03 version) to 1.11,"
@@ -21,7 +10,7 @@
    [Nn]*)
        ;;
    *)
-       ./loki_patch.$os h2_103_111.dat .
+       ./loki_patch h2_103_111.dat .
    ;;
 esac
 

