update StormMesa gcc makefile

diff -urNp StormMesa/AMIGA/llibs/egcs/Makefile~ StormMesa/AMIGA/llibs/egcs/Makefile
--- StormMesa/AMIGA/llibs/egcs/Makefile~	1998-10-03 02:16:48.000000000 +0300
+++ StormMesa/AMIGA/llibs/egcs/Makefile	2017-07-20 17:56:00.000000000 +0300
@@ -4,34 +4,30 @@ OBJ = $(OBJ1) $(OBJ2)
 
 # Modify the following two lines for your assembler
 # This version uses phxass from the StormC distribution
-ASM = StormC:StormSYS/PhxAss
+ASM = vasm
 
 # This line specifies the include directory
-ASM_OPTS = i=lvo
+ASM_OPTS = -Faout -phxass -quiet -Ilvo
 
-# The install program. Used to avoid conflics with the CBM install
-INSTALL = /usr/bin/install
+CC = m68k-amigaos-gcc
+AR = m68k-amigaos-ar
+RANLIB = m68k-amigaos-ranlib
 
-all: warning libgl.a
+all: libgl.a
 
 .s.o:
-	$(ASM) $(ASM_OPTS) $<
+	$(ASM) $(ASM_OPTS) -o $@ $<
 
-CFLAGS = -g -gstabs
+CFLAGS = -g -noixemul
+# -gstabs
+# -mcrt=clib2
 
 # If your assembler generates a.out format, you must change the lines
 # below
 libgl.a: $(OBJ)
-	-rm obj.*
-	hunk2aout $(OBJ1) >/dev/null
-	-rm -f libgl.a
-	ar cru libgl.a $(OBJ2) obj.*
-	ranlib libgl.a
-
-install:
-	$(INSTALL) --directory /usr/lib
-	ln -s /usr/lib/libgl.a /usr/lib/libglut.a
-	ln -s /usr/lib/libgl.a /usr/lib/libglu.a
+	$(AR) cru libgl.a $(OBJ2)  $(OBJ1)
+	$(RANLIB) libgl.a
+
+clean:
+	$(RM) *.a *.o
 
-warning:
-	@echo Make sure to read the README file
